File: LeetCode/Study Plan/Practice Problems/Medium/Convert Sorted List To Binary Search Tree/Go/convertSortedListToBST.go
package main

import "fmt"

type ListNode struct {
	Val  int
	Next *ListNode
}

func NewL(val int) *ListNode { return &ListNode{Val: val} }

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func New(val int) *TreeNode { return &TreeNode{Val: val} }

func sortedListToBST(head *ListNode) *TreeNode {
	sortedList := []int{}

	for head != nil {
		sortedList = append(sortedList, head.Val)
		head = head.Next
	}

	return bst(sortedList, 0, len(sortedList)-1)
}

func bst(arr []int, l, r int) *TreeNode {
	if l > r {
		return nil
	}

	mid := l + (r-l)/2
	root := New(arr[mid])
	root.Left = bst(arr, l, mid-1)
	root.Right = bst(arr, mid+1, r)
	return root
}

func main() {
	head := NewL(-10)
	head.Next = NewL(-3)
	head.Next.Next = NewL(0)
	head.Next.Next.Next = NewL(5)
	head.Next.Next.Next.Next = NewL(9)
	fmt.Println(sortedListToBST(head))
}

File: LeetCode/Medium/Convert Sorted List To Binary Search Tree/Java/convertSortedListToBST.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode sortedListToBST(ListNode head) {
        List<Integer> sortedList = new ArrayList<>();

        while(head != null) {
            sortedList.add(head.val);
            head = head.next;
        }

        return bst(sortedList, 0, sortedList.size()-1);
    }

    public TreeNode bst(List<Integer> arr, int l, int r) {
        if (l > r) return null;

        int mid = l + (r-l)/2;
        TreeNode root = new TreeNode(arr.get(mid));
        root.left = bst(arr, l, mid-1);
        root.right = bst(arr, mid+1, r);
        return root;
    }
}
